datafusion/sql/src/select.rs

```diff
@@ -440,11 +460,31 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
                 };
                 projection_exprs.extend(inner_projection_exprs);
 
+                let mut list_recursions = vec![];
+                let unnest_col_vec = unnest_columns
+                    .into_iter()
+                    .map(|(col, maybe_list_unnest)| {
+                        match maybe_list_unnest {
+                            None => {}
+                            Some(list_unnest) => {
+                                list_recursions.extend(list_unnest.into_iter().map(
+                                    |unnest_list| RecursionUnnestOption {
+                                        input_column: col.clone(),
+                                        output_column: unnest_list.output_column,
+                                        depth: unnest_list.depth,
+                                    },
+                                ));
+                            }
+                        }
```

Comment:
I think you could simplify this code (reduce recursion level) with `if let` like

```suggestion
                        if let Some(list_unnest) = maybe_list_unnest {
                                list_recursions.extend(list_unnest.into_iter().map(
                                    |unnest_list| RecursionUnnestOption {
                                        input_column: col.clone(),
                                        output_column: unnest_list.output_column,
                                        depth: unnest_list.depth,
                                    },
                                ));
                            }
                        }
```

Also I wonder if it would be possible / easier for `rewrite_recursive_unnests_bottom_up` to simply return a `HashSet<Column, RecursionUnnestOption>` directly rather than having to convert the results

Reply:
true, i can refactor this function so it returns 3 result:
- inner_projection_exprs
- unnest_placeholder_columns
- outer_projection_exprs